Additional query-level metrics that can be emitted as Logs

cjksplunk opened the pull request:
#### Description
Adding additional query-level metrics which can be emitted as Logs to include the query and plan text. Specifically:
- mysql.query.time.lock
- mysql.query.time.cpu
- mysql.query.rows.total
- mysql.query.rows.returned
- mysql.query.time.total
- mysql.query.calls

Includes option to emit as metrics, but will exclude the query and plan data. Note that there are already existing metrics to include some of this data, but it is gathered and emitted differently and not compatibly with these new metrics.

<!--Describe what testing was performed and which tests were added.-->
#### Testing
Existing tests were updated and run to include the new metrics gathered.

<!--Describe the documentation added.-->
#### Documentation
Updated ReadMe.md to include new params and requirements. The above metrics are reported as deltas between reporting cycles, and include only the most costly queries in terms of total execution time. It is possible (even likely) that some queries will consistently be listed that are not necessarily troublesome but simply take time to execute. Future enhancements may include filtering for such cases.
<!--Please delete paragraphs that you did not use before submitting.-->

Commit:
Additional query-level metrics that can be emitted as Logs to include query and plan text

Commit:
Updating tests

atoulme commented on receiver/mysqlreceiver/README.md:
```diff
@@ -47,6 +51,16 @@ The following settings are optional:
   - `time_limit` - maximum time from since the statements have been observed last time (default=`24h`)
   - `limit` - limit of records, which is maximum number of generated metrics (default=`250`)
 
+- `query_metrics_as_logs` (default = true): Some query metrics are emitted along with the full query text as logs. As described above with statement_events, there is a limit to teh length of attribute values in Metrics, but that limit does not apply to Logs. If this is enabled, the metric values are emitted as Log attribute values. Note that because of this, other Metric datapoint data, such as a ...
```
```suggestion
- `query_metrics_as_logs` (default = true): Some query metrics are emitted along with the full query text as logs. As described above with statement_events, there is a limit to the length of attribute values in Metrics, but that limit does not apply to Logs. If this is enabled, the metric values are emitted as Log attribute values. Note that because of this, other Metric datapoint data, such as attributes, will not be communicated.
```

atoulme replied:
this only applies if the receiver is on a log pipeline, and this is why this is defaulted to true, correct? What if it is set to false and used on a log pipeline, is the receiver then doing nothing?

cjksplunk replied:
This is inaccurate as I changed it to default to false. If this is false, then a separate function executes that emits the metrics through the metrics pipeline.

cjksplunk replied:
after discussion, I'll remove this switch and emit both metrics and logs

atoulme commented on receiver/mysqlreceiver/client.go:
```diff
@@ -243,8 +284,8 @@ func (c *mySQLClient) getVersion() (*version.Version, error) {
 	}
-	version, err := version.NewVersion(versionStr)
-	return version, err
+	vrsn, err := version.NewVersion(versionStr)
+	return vrsn, err
```
maybe can make this rename in a separate PR - or just return without assigning to variables at all

cjksplunk replied:
no problem

atoulme commented on receiver/mysqlreceiver/factory.go:
```diff
@@ -53,7 +56,7 @@ func createMetricsReceiver(
 
 	ns := newMySQLScraper(params, cfg)
-	s, err := scraper.NewMetrics(ns.scrape, scraper.WithStart(ns.start),
+	s, err := scraper.NewMetrics(ns.scrapeMetrics, scraper.WithStart(ns.start),
```
can we do this rename in a small PR first?

cjksplunk replied:
Can do.

atoulme commented on receiver/mysqlreceiver/go.mod:
```diff
@@ -1,6 +1,6 @@
 module github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mysqlreceiver
 
-go 1.23.0
+go 1.24.0
```
we use go 1.23 for now, as a repo policy

cjksplunk replied:
Noted, will adjust

cjksplunk commented on receiver/mysqlreceiver/README.md:
```diff
@@ -68,5 +82,5 @@ The full list of settings exposed for this receiver are documented in [config.go
 ## Metrics
 
-Details about the metrics produced by this receiver can be found in [metadata.yaml](./metadata.yaml)
+Details about the metrics produced by this receiver can be found in [metadata.yaml](./metadata.yaml). Note that the disabling of most metrics does not work.
```
That's a fact, actually :P. The implementation as it was when I came to it did not honor any of those settings, and aside from what I have done, still does not. It should be followed up on.

atoulme replied:
ok!

atoulme commented on receiver/mysqlreceiver/metadata.yaml:
```diff
@@ -3,7 +3,7 @@ type: mysql
   class: receiver
   stability:
-    beta: [metrics]
+    beta: [metrics, logs]
```
```suggestion
    beta: [metrics]
    development: [logs]
```

cjksplunk replied:
done

Commit:
reverted changes to var name

Commit:
removed query_metrics_as_logs

Commit:
reverted go to 1.23.0

Commit:
set new metrics to default enabled false

Commit:
updated to remove query_metrics_as_logs

Commit:
updated to remove query_metrics_as_logs

Commit:
moved logs to development stability

Commit:
Because metrics and logs are now both reported, but are not gathered concurrently, there is a need to track the last gathering time for each path. Also updated time calculation

Commit:
Merge branch 'main' into mysql-receiver-query-metrics

cjksplunk commented:
Moved to Draft per management until further internal review.

Commit:
Changing metric names and many log metrics per EO, adding scope name and version, updating test data.

Commit:
Removed now unused columns and edited appropriate query

Commit:
Updated config per EO to have top_query_collection.top_query_count

Commit:
Updated log EventName

Commit:
Moved some attribute keys to constants

Commit:
removed unnecessary empty log checking

Commit:
re-added some metrics to the query.
moved EXPLAIN queries to a transaction to avoid creating too many db connections.
Added logging of performance_schema_max_sql_text_length since queries that are longer than that value will be truncated and cannot be used to derive the explain plan.

Commit:
Added enablement to top_query_collection

Commit:
updated the expected data

Commit:
defaulted the TopQueryCollection enablement to false per OTel guidelines

Commit:
Updated docs to include top_query_collection

Commit:
re-added log metric attributes that had been previously backed out.
compacted the JSON query plan, updated placeholder values if plan is not available.

Commit:
Updated to cover code updates

Commit:
Merge branch 'upstream-main' into mysql-receiver-query-metrics

Commit:
Updates covering changes requested by EO and test updates

Commit:
Adding missing test data file

Commit:
Updating docs to list Logs attributes emitted in top_query_collection
